### ISSUE TITLE
Row height does not go back to normal after clearing grouping

### ISSUE THREAD

**Issue by pavish** [mathesar-foundation/mathesar#3768]
## Description
* Group a table by a column, notice that the group headers are displayed.
* Remove the grouping.
* Notice that the rows in the place on the now-removed headers (i.e. with the same index) retain the height of the group instead of going back to normal row height.
* When group is applied:
  ![Screenshot 2024-08-20 at 7 29 09 PM](https://github.com/user-attachments/assets/db011f41-c6fe-4a2e-90e0-378df5f21ad2)
* After group is removed:
  ![Screenshot 2024-08-20 at 7 29 21 PM](https://github.com/user-attachments/assets/0807a98e-ced9-42ab-aa0a-1aa9486f00b5)

**Comment by pavish** [comment]
This is a regression noticed after https://github.com/mathesar-foundation/mathesar/pull/3755.

**Comment by seancolsen** [comment]
I spent several hours troubleshooting this today but was unfortunately not able to identify the cause. And now I need to move on to higher-priority work. I'm commenting so that I (or someone else) can pick this back up at some point roughly where I left off. And perhaps a little rubber-duck debugging in public will also help...

- The bug is present for TableView when using VirtualList but not present without VirtualList (e.g. in the context of a table widget on the record page).
- The bug is present on develop but not in 0.1.7. It's not clear to me when it broke. Despite Pavish having noticed this in the course of reviewing #3755, I don't think that's the culprit. API incompatibilities and intentionally broken functionality make it difficult to verify.
- There appear to have been no meaningful changes within the VirtualList functionality since 0.1.7.
- There were some changes in `display.ts` within the derivation for `displayableRecords`, but I tried taking those changes out, and the bug persists.
- Very little of the code that seems relevant to this behavior seems to have changed. So I'm left with the hunch that something changed in a way that's incompatible with the relevant code. My hunch (very low confidence) is that the fix might end up being within some code that hasn't changed in years.
- The code that handles the height of rows within the VirtualList system is quite complex and not very readable. Much of it is third-party code that Pavish ported over a long time ago. I've spent some time trying to understand how it works, but it will require quite a bit of reading before I'm able to get my head around it enough.
- Within `Body.svelte` the value of `$displayableRecords` is updating and `getItemSizeFromIndex` gets called. When setting grouping, those events happen in the correct order — first `$displayableRecords` updates; then `getItemSizeFromIndex` is called. But when _un-setting_ grouping, they get called out of order so the stale rows are used when determining the row height.
- **The direction I'd like to head next** is the code inside `ScrollAndResetHandler.svelte`. There appears to be some imperative logic at `$: void resetIndex` which recalculates row height under certain conditions. My next steps are to understand that code and to poke at it to see if I can jostle something free that will help. I'm especially curious to understand the rationale uses of `tick()` within `resetIndex`.

**Comment by seancolsen** [comment]
I played with this again for a little bit today and want to capture my progress...

I was able to hack my way into getting the row height adapting correctly by doing the following:

1. In `ScrollAndResetHandler.svelte`, I removed `state`, `$state`, `prevRecordState`, and `_recordState`.
1. In `display.ts` I modified the derivation for `displayableRecords` so that it doesn't take `newRecords` into account.

Of course, these changes break other things! But they are helpful for the sake of troubleshooting. What I see happening in `develop` is this:

1. `displayableRecords` updates twice — once for changes to `savedRecordRowsWithGroupHeaders`, and again for changes to `newRecords`. Those are both updated when fetching records.
1. Each update triggers the `resetIndex` function to run, but the logic therein relies on comparing grouping to its previous value.

To me this feels like several different thinks are kind of messy and leading to a tricky reactivity pickle. I'm going to think on this a bit while chipping away at other work.
